### ISSUE TITLE
FSDP2 forward prefetch causes duplicate AllGather with reentrant HF activation checkpointing

### ISSUE THREAD

**Issue by DavidFFFan** [NVIDIA-NeMo/Automodel#3373]
# FSDP2 forward prefetch causes duplicate AllGather with reentrant HF activation checkpointing

## Summary

When a Hugging Face Transformers model is trained with FSDP2 activation
checkpointing and explicit FSDP2 forward prefetch enabled, the backward pass
performs duplicate parameter AllGather operations.

The issue occurs with the following combination:

- Hugging Face native gradient checkpointing
- `use_reentrant=True`
- FSDP2 parameter resharding
- `enable_fsdp2_prefetch=True`
- `fsdp2_forward_prefetch_depth > 0`

AutoModel currently explicitly enables the reentrant implementation here:

```python
model.gradient_checkpointing_enable(
    gradient_checkpointing_kwargs={"use_reentrant": True}
)
```

https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/components/distributed/parallelizer.py#L400

## Test setup

- World size: 8
- PyTorch 2.13.0
- Transformers 4.57.1
- Current NeMo AutoModel FSDP2 implementation
- A small 4-layer Hugging Face `LlamaForCausalLM`

All parameter gradients were present and finite in every test.

## Reproduction configuration

Equivalent distributed settings:

```yaml
distributed:
  activation_checkpointing: true
  activation_checkpointing_scope: all
  reshard_after_forward: true
  enable_fsdp2_prefetch: true
  fsdp2_backward_prefetch_depth: 2
  fsdp2_forward_prefetch_depth: 1
```

`reshard_after_forward: true` was used so every decoder layer requires a
parameter AllGather during backward, making the communication count
unambiguous.

The test instrumented the FSDP2 parameter-group `unshard()` path and counted
only operations that actually issued an AllGather. Calls that found parameters
already unsharded or with an outstanding unshard operation were not counted.

## Results

For a model with four FSDP-sharded decoder layers:

| Activation checkpointing | `use_reentrant` | Explicit prefetch | Backward depth | Forward depth | Forward AllGather | Backward AllGather |
|---|---:|---:|---:|---:|---:|---:|
| Disabled | N/A | Disabled | N/A | N/A | 5 | 4 |
| Enabled | `True` | Disabled | N/A | N/A | 5 | 4 |
| Enabled | `True` | Enabled | 1 | 1 | 5 | **7** |
| Enabled | `True` | Enabled | 2 | 1 | 5 | **7** |
| Enabled | `True` | Enabled | 2 | 0 | 5 | 4 |
| Enabled | `False` | Enabled | 2 | 1 | 5 | 4 |

The five forward AllGather operations correspond to the root FSDP unit plus
the four decoder-layer FSDP units.

Without explicit forward prefetch, activation recomputation does not increase
the number of real backward AllGather operations.

With reentrant checkpointing and forward prefetch depth 1, backward AllGather
increases from 4 to 7.

For `N` decoder layers that all reshard after forward, the observed pattern is:

```text
Expected:                         N
Reentrant AC + forward prefetch:  N + (N - 1) = 2N - 1
```

Changing backward prefetch depth from 1 to 2 does not change the duplicate
collective count.

Either of the following restores the expected count:

```text
fsdp2_forward_prefetch_depth = 0
```

or:

```text
use_reentrant = False
```

## Observed sequence

Without explicit forward prefetch, multiple hooks may attempt `unshard()`
during backward:

1. FSDP2 backward prefetch starts the layer's parameter AllGather.
2. Reentrant checkpointing reruns the layer forward and invokes its forward
   pre-hook.
3. The backward pre-hook may check the parameters again.

The latter attempts find the parameters already unsharded or with an
outstanding unshard operation, so they do not issue additional collectives.

The duplicate communication appears after explicit forward-prefetch chains are
installed.

Backward starts from the final decoder layer. During activation recomputation
of an earlier layer, its normal forward-prefetch chain runs again and fetches
the following layer even though that layer has completed backward and has been
resharded.

For example, with layers 0-3:

```text
backward-prefetch layer 3
backward-prefetch layer 2
recompute layer 2 -> forward-prefetch layer 3 again
backward-prefetch layer 1
recompute layer 1 -> forward-prefetch layer 2 again
backward-prefetch layer 0
recompute layer 0 -> forward-prefetch layer 1 again
```

This produces three redundant AllGather operations for four layers.

The explicit prefetch chains are configured here:

https://github.com/NVIDIA-NeMo/Automodel/blob/main/nemo_automodel/components/distributed/parallelizer.py#L1119

## Expected behavior

Activation recomputation should not cause the forward-prefetch chain to fetch
FSDP units that have already completed backward computation.

Each resharded decoder layer should require at most one parameter AllGather for
its backward computation.

## Workarounds

Disable explicit AutoModel FSDP2 prefetch:

```yaml
distributed:
  activation_checkpointing: true
  enable_fsdp2_prefetch: false
```

Alternatively, retain backward prefetch but disable forward prefetch:

```yaml
distributed:
  activation_checkpointing: true
  enable_fsdp2_prefetch: true
  fsdp2_backward_prefetch_depth: 2
  fsdp2_forward_prefetch_depth: 0
```

## Suggested fixes

Possible solutions include:

1. Use Hugging Face native gradient checkpointing with
   `use_reentrant=False` under FSDP2.
2. Do not install FSDP2 forward-prefetch chains when reentrant Hugging Face
   gradient checkpointing is enabled.
3. Suppress forward prefetch during activation recomputation.
4. Add a distributed regression test that counts parameter AllGather
   operations for FSDP2, activation checkpointing, and explicit prefetch.

The `use_reentrant=False` control test retained forward and backward prefetch
and produced the expected four backward AllGather operations.
